File: Sales_P3/script/2_budget sales.py
import pandas as pd
from pathlib import Path
from janitor import clean_names


# Path
path = Path(__file__).parent.parent


# Functions
def rename_columns(df):
    df = df.rename(
        columns={
            "Material (local)": "Product",
            "Profit Center": "Profit Ctr",
            "Ext./ICO": "Account Class",
            "Sold-to (central)": "Sold-to party",
            "Sold-to (central) (pivot)": "Sold-to Name 1",
        }
    )
    return df


def select_columns(df):
    col_category = [
        "Account Class",
        "Product",
        "Material Description",
        "Profit Ctr",
        "Sold-to party",
        "Sold-to Name 1",
    ]
    col_vol = [f"2023_{month:02d} vol" for month in range(1, 13)]
    col_amt = [f"2023_{month:02d} k LC" for month in range(1, 13)]

    df = df[col_category + col_vol + col_amt]
    return df


def process_textual_columns(df):
    data_types = {
        "Account Class": str,
        "Product": str,
        "Material Description": str,
        "Profit Ctr": str,
        "Sold-to party": str,
        "Sold-to Name 1": str,
    }
    df = df.astype(data_types)
    return df


def process_numeric_columns(df):
    col_category = [
        "Account Class",
        "Product",
        "Material Description",
        "Profit Ctr",
        "Sold-to party",
        "Sold-to Name 1",
    ]
    col_vol = [f"2023_{month:02d} vol" for month in range(1, 13)]
    col_amt = [f"2023_{month:02d} k LC" for month in range(1, 13)]

    # Split data into volume and sales amount dataframes
    df_vol = df[col_category + col_vol]
    df_amt = df[col_category + col_amt]

    # Pivot both dataframes to tidy format
    df_vol = pd.melt(df_vol, id_vars=col_category, var_name="Period", value_name="Qty")
    df_amt = pd.melt(
        df_amt, id_vars=col_category, var_name="Period", value_name="Sales_k_LC"
    )

    # Combine the two dataframes
    df = pd.concat([df_vol, df_amt["Sales_k_LC"]], axis="columns")

    # Clean up the 'Period' column by removing ' vol' from the values
    df["Period"] = df["Period"].str.replace(" vol", "")

    # Specify data types
    df = df.astype({"Qty": float, "Sales_k_LC": float})

    return df


def split_period(df):
    # Split Period into Year / Month
    df[["Year", "Month"]] = df["Period"].str.split("_", expand=True)
    df[["Year", "Month"]] = df[["Year", "Month"]].astype(float)
    # To match with Actual
    # df["Year"] = df["Year"] + 1
    # Specify data types
    df = df.astype({"Year": int, "Month": int})
    return df


def add_plant_info(row):
    # Assign Plnt based on Profit Ctr
    if row["Profit Ctr"] == "50802-018":
        return "2182"
    else:
        return "0180"


def add_account_class(row):
    # Change Account Class to match with actual data
    if row["Account Class"] == "ICO":
        return "NSI"
    elif row["Account Class"] == "Ext." and "Continental" in row["Sold-to Name 1"]:
        return "NSR"
    else:
        return "NSE"


def add_record_type(row):
    # Add a new column RecordType
    if row["Product"] == "0":
        return "B"
    else:
        return "F"


def sales_in_full(df):
    df["Sales_LC"] = df["Sales_k_LC"] * 1000
    return df


def remove_missing_values(df):
    # Remove NA values in Profit center
    return df.dropna(subset=["Profit Ctr"])


def remove_zero_values(df):
    # Remove zero values in Qty and Sales
    return df[(df["Qty"] != 0) | (df["Sales_LC"] != 0)]


def reorder_columns(df):
    df = df[
        [
            "Period",
            "Year",
            "Month",
            "Profit Ctr",
            "Plnt",
            "Account Class",
            "RecordType",
            "Product",
            "Material Description",
            "Sold-to party",
            "Sold-to Name 1",
            "Qty",
            "Sales_LC",
        ]
    ]
    return df


def clean_column_names(df):
    df = df.clean_names()
    df.columns = df.columns.str.strip("_")
    return df


def main():

    # Filenames
    input_file = (
        path
        / "data"
        / "Plan"
        / "2023_BPR_Consolidation_20220907_CMU Material EQ update.xlsx"
    )
    output_file = path / "output" / "2_budget_sales.csv"

    # Read data
    df = pd.read_excel(input_file, sheet_name="Cons", skiprows=2)

    # Process data
    df = (
        df.pipe(rename_columns)
        .pipe(select_columns)
        .pipe(process_textual_columns)
        .pipe(process_numeric_columns)
        .pipe(split_period)
    )

    # Add columns
    df["Plnt"] = df.apply(add_plant_info, axis="columns")
    df["Account Class"] = df.apply(add_account_class, axis="columns")
    df["RecordType"] = df.apply(add_record_type, axis="columns")

    # Miscellaneous
    df = (
        df.pipe(sales_in_full)
        .pipe(remove_missing_values)
        .pipe(remove_zero_values)
        .pipe(reorder_columns)
        .pipe(clean_column_names)
    )

    # Write data
    df.to_csv(output_file, index=False)
    print("A file is created")


if __name__ == "__main__":
    main()

File: Sales_P3/src/dplyr/2_budget_sales.R
library(dplyr)
library(stringr)
library(readxl)
library(readr)
library(janitor)
library(here)
library(glue)


# Path
path <- here("Sales_P3")


# Functions
rename_columns <- function(df) {
  df <- df |>
    rename(c(
      "Product" = "Material (local)",
      "Profit Ctr" = "Profit Center",
      "Account Class" = "Ext./ICO",
      "Sold-to party" = "Sold-to (central)",
      "Sold-to Name 1" = "Sold-to (central) (pivot)"
    ))
  return(df)
}


select_columns <- function(df, year) {
  df <- df |>
    select(c(
      "Account Class",
      "Product",
      "Material Description",
      "Profit Ctr",
      "Sold-to party",
      "Sold-to Name 1",
      glue("{year}_01 vol"):glue("{year}_12 vol"),
      glue("{year}_01 k LC"):glue("{year}_12 k LC")
    ))
  return(df)
}


process_volume_and_amount <- function(df) {
  # Split data into Volume and Sales amount
  df_vol <- df |>
    select(c(1:"Sold-to Name 1", ends_with("vol")))

  df_amt <- df |>
    select(c(1:"Sold-to Name 1", ends_with("k LC")))

  # Tidyr::Pivot_longer
  df_vol <- df_vol |>
    tidyr::pivot_longer(
      cols = ends_with("vol"),
      names_to = "Period",
      values_to = "Qty",
    )

  df_amt <- df_amt |>
    tidyr::pivot_longer(
      cols = ends_with("k LC"),
      names_to = "Period",
      values_to = "Sales_k_LC",
    )

  # Combine data with Volume and Sales amount
  df <- bind_cols(
    df_vol, df_amt |> select("Sales_k_LC")
  ) |>
    mutate(
      Period = str_remove(.data$Period, " vol")
    )

  return(df)
}


process_miscellaneous <- function(df) {
  # To match with actual data, etc ----
  df <- df |>
    mutate(
      Sales_LC = df$Sales_k_LC * 1000,
      Plnt = case_when(
        df["Profit Ctr"] == "50802-018" ~ "2182",
        TRUE ~ "0180"
      )
    ) |>
    select(!c("Sales_k_LC")) |>
    select(c(
      "Period", "Profit Ctr", "Account Class",
      "Plnt", "Product", "Material Description",
      "Sold-to party", "Sold-to Name 1",
      "Qty", "Sales_LC"
    ))
  return(df)
}


split_period <- function(df) {
  # Split Period into Year / Month
  df <- df |>
    tidyr::separate(
      "Period",
      into = c("Year", "Month"),
      sep = "_"
    ) |>
    mutate(across(c("Year":"Month"), as.double))
  return(df)
}


arrange_by_month <- function(df) {
  df <- df |>
    arrange(.data$Month)
  return(df)
}


add_account_class <- function(df) {
  # Change Account Class to match with actual data
  df <- df |>
    mutate(`Account Class` = case_when(
      df["Account Class"] == "ICO" ~ "NSI",
      df["Account Class"] == "Ext." &
        str_detect(df["Sold-to Name 1"], "Continental") ~ "NSR",
      TRUE ~ "NSE"
    ))
  return(df)
}


add_record_type <- function(df) {
  # Add a new column RecordType
  df <- df |>
    mutate(
      RecordType = case_when(
        Product == "0" | is.na(Product) ~ "B",
        TRUE ~ "F"
      ),
      .after = "Profit Ctr"
    )
  return(df)
}

remove_na <- function(df) {
  # Remove NA values in Profit center
  df <- df |>
    filter(!is.na(df["Profit Ctr"]))
  return(df)
}


remove_zero <- function(df) {
  # Remove zero values in Qty and Sales
  df <- df |>
    filter(!(df$Qty == 0 & df$Sales_LC == 0))
  return(df)
}


main <- function() {
  # Variable
  year <- 2024

  # Filenames
  input_file <- here(
    path, "data", "plan",
    glue(
      "{year}_Consolidate_Sales_Budget 2024 CDP BIN_version_Incl.ICO_231004.xlsx"
    )
  )
  output_file <- here(path, "output", "2_budget_sales.csv")

  # Read data
  df <- read_excel(input_file, sheet = "Cons", skip = 2)

  # Process data
  df <- df |>
    rename_columns() |>
    select_columns(year) |>
    process_volume_and_amount() |>
    process_miscellaneous() |>
    split_period()

  df <- df |>
    arrange_by_month() |>
    add_account_class() |>
    add_record_type()

  df <- df |>
    remove_na() |>
    remove_zero() |>
    clean_names()

  # Write data
  write_csv(df, output_file)
  print("A file is created")
}

main()
